src/bootloaders/systemd-class.c

```diff
@@ -384,6 +384,131 @@ bool sd_class_set_default_kernel(const BootManager *manager, const Kernel *kerne
         return true;
 }
 
+static char *parse_kernel_from_loader(char *conf, const BootManager *manager)
+{
+        char ktype[32] = { 0 };
+        char version[16] = { 0 };
+        long int release = -1;
+        const char *prefix = NULL;
+        char * const hdr = "default ";
+        char *start = NULL;
+        char *end = NULL;
+        int field = 0;
+        size_t ln;
+
+        if (!conf) {
+                return NULL;
+        }
+
+        /*
+         * Looking for the 'default $kernel' line, it may contain extra
+         * lines around it, but just grab the start and end of the
+         * $kernel content.
+         */
+        start = strstr(conf, hdr);
+        if (!start) {
+                return NULL;
+        }
+        start = start + strlen(hdr);
+        *(start - 1) = '-';
+        end = strstr(start, "\n");
+
+        /* Make end a nul character */
+        if (end) {
+                *end = 0;
+        } else {
+                end = start + strlen(start);
+        }
+
+        /*
+         * Horrible manually parsing a string of the form:
+         * 'Clear-linux-pk414-standard-4.14.74-115' below.
+         *
+         * Idea is to parse backwards using the dashes '-' as field
+         * boundries. First comes release #, then kernel version and
+         * finally the kernel identifier with the vendor prefix.
+         *
+         * So given the example we want:
+         * release = 115;
+         * version = "4.14.74";
+         * ktype = "pk414-standard";
+         */
+        for (char *c = end; c != start; c--) {
```

Comment:
Nit: I think it's generally safer to use `c >= start` in case it would occur to increment the counter in parsing (later).

Reply:
I agree and in this case start is already modified so going one past the start and comparing pointers wouldn't be undefined though I really don't want this code to ever play with the c pointer location in the body X(.

Reply:
Ah it needs to be c <= start though in this case since it is going backwards no?

Reply:
`c > start`. the cycle starts from `end` backwards. it's ok to hit `*start` itself too since you know it's there and at this point you'd fail to parse anyways (the real exit condition is under `if (field == 2)`).

Reply:
Gah yes, thanks.